### PR TITLE
Deployment for EFK to support job history

### PR DESCRIPTION
**Pull request by doubleyoung18**
deploy efk on k8s

### CODE REVIEW

**Comment by yqwang-ms** [comment]
@wangdian @ydye Please help to check

File: src/fluentd/build/fluent.conf [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
+# BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+# DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+<source>
+  @type  forward
```
Comment (yqwang-ms) [inline src/fluentd/build/fluent.conf]: Why not use fluentd.conf

File: src/elasticsearch/deploy/elasticsearch.yaml [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+      app: elasticsearch
+  template:
+    metadata:
+      labels:
+        app: elasticsearch
+        role: master
+    spec:
+      containers:
+      - name: elasticsearch-master
+        image: elasticsearch:7.3.0
```
Comment (ydye) [inline src/elasticsearch/deploy/elasticsearch.yaml]: Please refer to prometheus. Don't directly pull the image from official repo. Suggest to do a build steps.
https://github.com/microsoft/pai/blob/master/src/prometheus/build/prometheus.dockerfile

File: src/elasticsearch/deploy/elasticsearch.yaml [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        resources:
```
Comment (ydye) [inline src/elasticsearch/deploy/elasticsearch.yaml]: For resource requirement, please add system qos flag. 
https://github.com/microsoft/pai/blob/master/src/prometheus/deploy/prometheus-deployment.yaml.template#L52

File: src/fluentd/deploy/fluentd.yaml [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    spec:
+      containers:
+      - name: fluentd-es
+        # need to build fluentd image with elasticsearch plugin
+        # current use cnych/fluentd-elasticsearch:v2.0.4
+        image: cnych/fluentd-elasticsearch:v2.0.4
+        env:
+        - name: FLUENTD_ARGS
+          value: --no-supervisor -q
+        resources:
```
Comment (ydye) [inline src/fluentd/deploy/fluentd.yaml]: Resource reqiruement should have a cluster level flag.

File: src/fluentd/deploy/fluentd.yaml [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+          mountPath: /var/log
+        - name: varlibdockercontainers
+          mountPath: /var/lib/docker/containers
+          readOnly: true
+        - name: config-volume
+          mountPath: /etc/fluent/config.d
+      volumes:
+      - name: varlog
+        hostPath:
+          path: /var/log
```
Comment (ydye) [inline src/fluentd/deploy/fluentd.yaml]: ```/var/log/fluentd``` will be better

File: src/fluentd/deploy/fluentd.yaml [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+      - name: varlibdockercontainers
+        hostPath:
+          path: /var/lib/docker/containers
```
Comment (ydye) [inline src/fluentd/deploy/fluentd.yaml]: This default value of ```docker data-root```. However, this value is configurable, and actually we have modified a lot of host to switch the path to other path.

File: src/kibana/deploy/kibana.yaml [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+            memory: 256Mi
+        ports:
+        - containerPort: 5601
+          protocol: TCP
+        env:
+        - name: "SERVER_NAME"
+          value: "kibana"
+        - name: "ELASTICSEARCH_HOSTS"
+          # TODO connect to ${elasticsearch_host: elasticsearch-service_port}
+          value: "http://10.151.40.4:30100"
```
Comment (ydye) [inline src/kibana/deploy/kibana.yaml]: You should change this value, and I think this address should be master ip?

File: src/kibana/deploy/kibana.yaml [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+      app: kibana
+  template:
+    metadata:
+      labels:
+        app: kibana
+    spec:
+      containers:
+      - name: kibana
+        image: kibana:7.3.0
+        resources:
```
Comment (ydye) [inline src/kibana/deploy/kibana.yaml]: Resource requirement should have a cluster level flag to judge whether enable the qos policy

File: src/kibana/deploy/kibana.yaml [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    matchLabels:
+      app: kibana
+  template:
+    metadata:
+      labels:
+        app: kibana
+    spec:
+      containers:
+      - name: kibana
+        image: kibana:7.3.0
```
Comment (ydye) [inline src/kibana/deploy/kibana.yaml]: Same as above.

File: src/fluentd/deploy/fluentd.yaml [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+      # supports critical pod annotation based priority scheme.
+      # Note that this does not guarantee admission on the nodes (#40573).
+      annotations:
+        scheduler.alpha.kubernetes.io/critical-pod: ''
+    spec:
+      containers:
+      - name: fluentd-es
+        # need to build fluentd image with elasticsearch plugin
+        # current use cnych/fluentd-elasticsearch:v2.0.4
+        image: cnych/fluentd-elasticsearch:v2.0.4
```
Comment (ydye) [inline src/fluentd/deploy/fluentd.yaml]: Same as above

File: src/fluentd/deploy/fluentd-config.yaml [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+
+    # Output to elasticsearch
+    <match frameworkcontroller.log>
+      @type copy
+      # index by kind
+      <store>
+        @type elasticsearch_dynamic
+        type_name _doc
+        # elasticsearch_host
+        host 10.151.40.4
```
Comment (ydye) [inline src/fluentd/deploy/fluentd-config.yaml]: harcode value. You should set this as a template value.

File: src/fluentd/deploy/service.yaml [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+cluster-type:
+  - k8s
+
+prerequisite:
+  - cluster-configuration
+
+template-list:
+  - fluentd.yaml
```
Comment (ydye) [inline src/fluentd/deploy/service.yaml]: remove this line

File: src/kibana/deploy/kibana.yaml [hunk 13]
```diff
@@ -0,0 +1,56 @@
+# Copyright (c) Microsoft Corporation
```
Comment (ydye) [inline src/kibana/deploy/kibana.yaml]: rename this file to kibana.yaml.template

File: src/fluentd/deploy/fluentd.yaml (hunk 6 above)
Comment (yqwang-ms) [reply src/fluentd/deploy/fluentd.yaml]: We adopt offical config:
https://github.com/fluent/fluentd-kubernetes-daemonset/blob/c40a8518ad1865c61786feef29fdd1acfaca956f/fluentd-daemonset-elasticsearch.yaml#L51-L56

**Pull request merged**